src/som/interpreter/Interpreter.java

```diff
@@ -25,297 +25,291 @@
   }
 
-  private void doReturnNonLocal() {
+  private SAbstractObject doReturnNonLocal(final Frame frame) throws ReturnException {
     // Handle the RETURN NON LOCAL bytecode
-    SAbstractObject result = getFrame().pop();
+    SAbstractObject result = frame.pop();
 
     // Compute the context for the non-local return
-    Frame context = getFrame().getOuterContext(universe.nilObject);
+    Frame context = frame.getOuterContext(universe.nilObject);
 
     // Make sure the block context is still on the stack
     if (!context.hasPreviousFrame(universe.nilObject)) {
       // Try to recover by sending 'escapedBlock:' to the sending object
       // this can get a bit nasty when using nested blocks. In this case
       // the "sender" will be the surrounding block and not the object
       // that actually sent the 'value' message.
-      SBlock block = (SBlock) getFrame().getArgument(0, 0);
+      SBlock block = (SBlock) frame.getArgument(0, 0);
       SAbstractObject sender =
-          getFrame().getPreviousFrame().getOuterContext(universe.nilObject).getArgument(0, 0);
-
-      // pop the frame of the currently executing block...
-      popFrame();
-
-      // pop old arguments from stack
-      SMethod method = getFrame().getMethod();
-      int numArgs = method.getNumberOfArguments();
-      for (int i = 0; i < numArgs; i += 1) {
-        getFrame().pop();
-      }
+          frame.getPreviousFrame().getOuterContext(universe.nilObject).getArgument(0, 0);
 
+      // TODO - check the frame and its stack
       // ... and execute the escapedBlock message instead
-      sender.sendEscapedBlock(block, universe, this);
-      return;
+      sender.sendEscapedBlock(block, universe, this, frame);
+      return frame.pop();
     }
 
-    // Unwind the frames
-    while (getFrame() != context) {
-      popFrame();
-    }
-
-    // Pop the top frame and push the result
-    popFrameAndPushResult(result);
+    // throw the exception to pass around the context and pop the right frames
+    throw new ReturnException(result, context);
   }
 
-  private void doSend(final int bytecodeIndex) {
-    // Handle the SEND bytecode
-    SSymbol signature = (SSymbol) getMethod().getConstant(bytecodeIndex);
-
-    // Get the number of arguments from the signature
+  private void doSend(final int bytecodeIndex, final Frame frame, final SMethod method) {
+    SSymbol signature = (SSymbol) method.getConstant(bytecodeIndex);
     int numberOfArguments = signature.getNumberOfSignatureArguments();
+    SAbstractObject receiver = frame.getStackElement(numberOfArguments - 1);
 
-    // Get the receiver from the stack
-    SAbstractObject receiver = getFrame().getStackElement(numberOfArguments - 1);
+    ValueProfile receiverClassValueProfile = method.getValueProfile(bytecodeIndex);
+    if (receiverClassValueProfile == null) {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
+      receiverClassValueProfile = receiver.getValueProfile();
```

Comment:
I am not convinced this is a good idea, but now I see what you are doing.
I would recommend against this, and simply construct a profile here at this point.